### PR TITLE
feat: Add new ptn modules `avm/ptn/azd/acr-container-app`

### LINKED ISSUE #1221: [Module Proposal]: `avm/ptn/azd/acr-container-app`
### Check for previous/existing GitHub issues/module proposals

- [X] I have checked for previous/existing GitHub issues/module proposals.

### Check this module doesn't already exist in the module indexes

- [X] I have checked for that this module doesn't already exist in the module indexes.

### Bicep or Terraform?

Bicep

### Module Classification?

Pattern Module

### Module Name

avm/ptn/azd/acr-container-app

### Module Details

Please add a Pattern Module named `avm/ptn/azd/acr-container-app`, this module will create the following two resources:

1. Container App.
2. Azure Container Registry (ACR).

For details, please refer to azd: 
https://github.com/Azure/azure-dev/blob/main/templates/common/infra/bicep/core/host/container-app.bicep

CC: @jongio

### Do you want to be the owner of this module?

No

### Module Owner's GitHub Username (handle)

jongio

### (Optional) Secondary Module Owner's GitHub Username (handle)

_No response_

### PR DESCRIPTION
**Pull request by NanaXiong00**
## Description
Fixes https://github.com/Azure/Azure-Verified-Modules/issues/1221

## Pipeline Reference

| Pipeline |
| -------- |
|     [avm.ptn.azd.acr-container-app](https://github.com/NanaXiong00/bicep-registry-modules/actions/workflows/avm.ptn.azd.acr-container-app.yml)    |

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

@jongio for notification.

### CODE REVIEW

File: avm/ptn/azd/container-app/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/azd/container-app/main.bicep:113]:
```suggestion
@description('Optional. The resource id of the user-assigned identity.')
param userAssignedIdentityResourceId string = ''
```

File: avm/ptn/azd/container-app/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/azd/container-app/main.bicep:53]: Could those contain secrets?

File: avm/ptn/azd/container-app/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/azd/container-app/main.bicep:6]:
```suggestion
param name string
```

The ptn itself is called 'Container-App'. As such it should ideally not be necessary to also have it mentioned in the name parameter of the module

File: avm/ptn/azd/container-app/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/azd/container-app/main.bicep:240]:
```suggestion
type environmentType = {
```

Small suggestion - but would be great as it aligns with the other modules

**Comment by NanaXiong00** [comment]
@AlexanderSehr I have updated based on your comments, please re-review.

**Review by AlexanderSehr — DISMISSED** [review]

File: avm/ptn/azd/container-app/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/azd/container-app/main.bicep:1]: The PR looks good to go @NanaXiong00.
I'd just have one question: Do you know why we happen to have 2 patterns that are named always the same?
```bicep
metadata name = 'avm/ptn/azd/container-apps'
metadata description = 'Creates an Azure Container Registry and an Azure Container Apps environment.'
```
and
```bicep
metadata name = 'avm/ptn/azd/container-app'
metadata description = 'Creates a container app in an Azure Container App environment.'
```

Namewise - I can't help but wonder if it's not mildly confusing 😄 I assume there's a common team responsible for the `azd` modules, so I'd just be curious to know if you feel like these 2 modules are, name- & description-wise, distinct enough.

Want to make sure we got this covered before either of the two are merged. ([ref](https://github.com/Azure/bicep-registry-modules/blob/2eeb04162c7f26fddf66bf5b7b401895c85cf6fe/avm/ptn/azd/container-apps/main.bicep))

Comment (NanaXiong00) [reply avm/ptn/azd/container-app/main.bicep:1]: @AlexanderSehr The above two ptn module names are named according to the two files [container-apps.bicep](https://github.com/Azure/azure-dev/blob/main/templates/common/infra/bicep/core/host/container-apps.bicep) and [container-app.bicep](https://github.com/Azure/azure-dev/blob/main/templates/common/infra/bicep/core/host/container-app.bicep) in azure-dev. These names have been reviewed and confirmed by @jongio.

Comment (AlexanderSehr) [reply avm/ptn/azd/container-app/main.bicep:1]: Fair and maybe it's just me that would consider this naming very very 'unfortunate'. @jongio, any guess as to why this happened? 
Wouldn't a name like `/azd/registry-linked-container-app` made more sense for this pattern, and `/azd/container-app-environment` for the other?

I'm not going to block your PRs over it, but given that these are published to the public registry for everybody to use, I at least wanted to bring it up.

Comment (NanaXiong00) [reply avm/ptn/azd/container-app/main.bicep:1]: @jongio What do you think of AlexanderSehr's suggestion above?

Comment (jongio) [reply avm/ptn/azd/container-app/main.bicep:1]: @wbreza - Thoughts on if we should use this opportunity to rename them?

Comment (AlexanderSehr) [reply avm/ptn/azd/container-app/main.bicep:1]: If we want to do it, we should do it now. Technically nothing is published and just the entry in the MAR file would need to be replaced & the GitHub teams would need to be renamed.

Comment (eriqua) [reply avm/ptn/azd/container-app/main.bicep:1]: Hey @jongio @NanaXiong00 did you have the chance to discuss the above renaming suggestion?

Comment (v-xuto) [reply avm/ptn/azd/container-app/main.bicep:1]: > Hey @jongio @NanaXiong00 did you have the chance to discuss the above renaming suggestion?

@jongio About the renaming suggestion, any ideas?

Comment (wbreza) [reply avm/ptn/azd/container-app/main.bicep:1]: Agree that now is the time to rename them since they are so similar.  The `container-apps` module is really all about creating the prereqs and dependencies required by the actual container apps.  Perhaps we should consider using one of the following:

- container-apps-stack
- container-apps-setup
- container-apps-foundation
- container-apps-prereqs

Comment (AlexanderSehr) [reply avm/ptn/azd/container-app/main.bicep:1]: @matebarabas, any thoughts? This would require us to 
- update the AVM issues for the modules
- create a new PR to the MAR file, updating the original values
- updating the PR(s) to reflect the new telemetry value, folder naming etc.

Just simply updating the name for both the 
- azd/container-app and
- azd/container-apps
modules as per whatever name is prefered for either one.

Not sure who wants to ultimately decide the name, but I think @wbreza's suggestions sound already a lot better :)

Comment (matebarabas) [reply avm/ptn/azd/container-app/main.bicep:1]: I'm fully supportive of renaming modules that don't have a meaningful enough name. In fact, I've already called out exactly this problem in the internal planning thread, but that conversation concluded on keeping the original names for those modules that were eventually (two of them got ditched in the process).

So, please let me know what's the new name for which one of these, and I'm happy to take the necessary steps (most of them highlighted by Alex above):
- `avm/ptn/azd/container-app` ---> to be renamed to ??? OR are we keeping this and just renaming the below one?
- `avm/ptn/azd/container-apps` ---> to be renamed to ???

Internally, I've also asked about this one:
- `avm/ptn/azd/container-app-upsert`---> to be renamed to ???

Thanks!

Comment (AlexanderSehr) [reply avm/ptn/azd/container-app/main.bicep:1]: I'd strongly recommend to rename both, as neither describes the corresponding module's content. Moreover, both are essentially just called the same which doesn't help.

Comment (eriqua) [reply avm/ptn/azd/container-app/main.bicep:1]: 100% agree with Alex here.

**Comment by ReneHezser** [comment]
@NanaXiong00 
We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.

Please reach out if any support is needed.

File: avm/ptn/azd/container-app/main.bicep
Comment (matebarabas) [reply avm/ptn/azd/container-app/main.bicep:1]: Thanks, @wbreza! Do you need any other input or approval from other folks in the team, or shall we continue with the renaming to the above? Thanks.
CC: @jongio

Comment (jongio) [reply avm/ptn/azd/container-app/main.bicep:1]: ship it

Comment (matebarabas) [reply avm/ptn/azd/container-app/main.bicep:1]: Excellent! Let me take the next steps. I'll let everyone know (here), when everything is ready.

Comment (matebarabas) [reply avm/ptn/azd/container-app/main.bicep:1]: Checklist:
- [x] update the AVM module proposal issues and BRM PRs with the new module names
- [x] update the module indexes with the new names
- [x] submit a PR to update the MAR manifest file with the new names
- [ ] MAR manifest PR merge confirmed (**pending, module PRs cannot be published until this step is complete!**)
- [ ] rename the related GH teams @jongio to grant me maintainer access on each of the below teams, so I can help:
  - [avm-ptn-azd-containerapp-module-contributors-bicep · Azure Team (github.com)](https://github.com/orgs/Azure/teams/avm-ptn-azd-containerapp-module-contributors-bicep)
  - [avm-ptn-azd-containerapp-module-owners-bicep · Azure Team (github.com)](https://github.com/orgs/Azure/teams/avm-ptn-azd-containerapp-module-owners-bicep)
  - [avm-ptn-azd-containerapps-module-contributors-bicep · Azure Team (github.com)](https://github.com/orgs/Azure/teams/avm-ptn-azd-containerapps-module-contributors-bicep)
  - [avm-ptn-azd-containerapps-module-owners-bicep · Azure Team (github.com)](https://github.com/orgs/Azure/teams/avm-ptn-azd-containerapps-module-owners-bicep)
- [ ] update both PR(s) to reflect the new telemetry value, folder naming etc.
  - [ ] https://github.com/Azure/bicep-registry-modules/pull/3234
  - [ ] https://github.com/Azure/bicep-registry-modules/pull/3221

**Comment by NanaXiong00** [comment]
@AlexanderSehr and @ReneHezser I have updated based on rename comments, please re-review.

File: avm/ptn/azd/container-app/main.bicep
Comment (jongio) [reply avm/ptn/azd/container-app/main.bicep]: Done

Comment (matebarabas) [reply avm/ptn/azd/container-app/main.bicep]: @jongio, you only added me as a member in each, can you please promote me to a maintainer so that I have rights to make changes? Thanks!

You need to check the box next to my name and then select change role: 
image

Then, select maintainer:
image

Comment (matebarabas) [reply avm/ptn/azd/container-app/main.bicep]: Thanks, Jon, for granting me access. I'm done with the edits and have removed myself from the teams.

**Comment by NanaXiong00** [comment]
@ReneHezser I have updated branch and run the Set-AVMModule utility, please review.

File: avm/ptn/azd/acr-container-app/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/azd/acr-container-app/main.bicep:147]: Misses passthru of `enableTelemetry` parameter

File: avm/ptn/azd/acr-container-app/modules/registry-access.bicep
Comment (AlexanderSehr) [inline avm/ptn/azd/acr-container-app/modules/registry-access.bicep:18]: Misses passthru of `enableTelemetry` parameter

File: .github/CODEOWNERS
Comment (AlexanderSehr) [inline .github/CODEOWNERS:15]: team does not exist, or not in the correct hierachy

File: avm/ptn/azd/acr-container-app/main.bicep
Comment (AlexanderSehr) [reply avm/ptn/azd/acr-container-app/main.bicep:147]: Hey @NanaXiong00, the fix was unfortunately not correct. It's about adding a new parameter to each module deployment's param block. **Not** a condition.
Also please don't self-resolve comments but just comment them once addressed. Makes re-reviews easier

File: avm/ptn/azd/acr-container-app/modules/registry-access.bicep
Comment (AlexanderSehr) [reply avm/ptn/azd/acr-container-app/modules/registry-access.bicep:21]: See other comment

File: .github/CODEOWNERS
Comment (AlexanderSehr) [reply .github/CODEOWNERS:15]: Hey @NanaXiong00, it's about this [spec](https://azure.github.io/Azure-Verified-Modules/specs/shared/#team-memberships) in the contribution guide.
Adding the correct parent team to your teams ensures you have the permissions in the future to approve PRs for your module

File: avm/ptn/azd/acr-container-app/main.bicep
Comment (NanaXiong00) [reply avm/ptn/azd/acr-container-app/main.bicep:147]: @AlexanderSehr I have updated by your comments.

File: avm/ptn/azd/acr-container-app/modules/registry-access.bicep
Comment (NanaXiong00) [reply avm/ptn/azd/acr-container-app/modules/registry-access.bicep:21]: @AlexanderSehr I have updated by your comments.

File: avm/ptn/azd/acr-container-app/main.bicep
Comment (eriqua) [inline avm/ptn/azd/acr-container-app/main.bicep:1]: This will be used as the readme title. I'd suggest a more user friendly short name, possibly starting with `Azd`

File: .github/CODEOWNERS
Comment (AlexanderSehr) [reply .github/CODEOWNERS:15]: Seems to be in order now 💪

File: avm/ptn/azd/container-app/main.bicep
Comment (AlexanderSehr) [reply avm/ptn/azd/container-app/main.bicep]: Resolving

File: avm/ptn/azd/acr-container-app/main.bicep
Comment (NanaXiong00) [reply avm/ptn/azd/acr-container-app/main.bicep]: @eriqua I have updated by your comments.

**Review by AlexanderSehr — APPROVED** [review]
Off we go @NanaXiong00, thank you for the great work 💪

**Pull request merged**
